cylc/flow/network/resolvers.py

```diff
@@ -153,47 +167,52 @@ def __init__(self, schd):
         w_ids = set()
         for nat_id in nat_ids:
             o_name, w_name, _ = nat_id.split(ID_DELIM, 2)
-            w_ids.add(f'{o_name}{ID_DELIM}{w_name}')
-        if self.schd.ws_data_mgr.workflow.id not in w_ids:
-            return []
-        flow_msg = self.schd.ws_data_mgr.get_entire_workflow()
+            flow_id = f'{o_name}{ID_DELIM}{w_name}'
+            if flow_id in self.data:
+                w_ids.add(flow_id)
         if node_type == 'proxy_nodes':
-            nodes = (list(getattr(flow_msg, 'task_proxies', []))
-                     + list(getattr(flow_msg, 'family_proxies', [])))
-        else:
-            nodes = list(getattr(flow_msg, node_type, []))
+            return sort_elements(
+                [node
+                 for flow in [self.data[w_id] for w_id in w_ids]
+                 for node in (
+                     [flow['task_proxies'][n_id]
+                      for n_id in nat_ids
+                      if n_id in flow['task_proxies']] +
+                     [flow['family_proxies'][n_id]
+                      for n_id in nat_ids
+                      if n_id in flow['family_proxies']])
+                 if node_filter(node, args)],
+                args)
         return sort_elements(
             [node
-             for node in nodes
-             if node.id in nat_ids and node_filter(node, args)],
+             for flow in [self.data[w_id] for w_id in w_ids]
+             for node in [flow[node_type][n_id]
+                          for n_id in nat_ids
+                          if n_id in flow[node_type]]
+             if node_filter(node, args)],
```

Comment:
Any chance of refactoring the block between L174 and L192 so it can be easier to understand?

Reply:
List comprehensions aren't the easiest to follow.. Would you like some comments instead? or the
```
 [flow['task_proxies'][n_id]
  for n_id in nat_ids
  if n_id in flow['task_proxies']]
```
block made a function?

Reply:
Fantastic if you can make the block a function - so we don't end up with so many of these list comprehensions (that are difficult to read at least for me).

Reply:
Moving the `node_type` logic out of the list comprehension it becomes pretty clear:

```python
if node_type == PROXY_NODES:
    node_types = [TASK_PROXIES, FAMILY_PROXIES]
else:
    node_types = [node_type]

return sort_elements(
    [
        node
        for flow in get_flow_data(self.data, w_ids)
        for node_type in node_types
        for node in get_data_elements(flow, nat_ids, node_type)
        if node_filter(node, args)
    ],
    args
)
```

Reply:
I guess I'm just not blessed with the talent of reading a complex list comprehension like this one with 3 for-loops and an if-block...

Reply:
Done.